File: packages/contracts/python/demodirector_contracts/editorial.py
from __future__ import annotations

from datetime import datetime
from typing import Literal

from pydantic import Field, model_validator

from demodirector_contracts.models import ContractModel, NonEmptyString

EditorialTemplateId = Literal[
    "hook",
    "framed_product",
    "feature_callout",
    "split_explanation",
    "proof_safety",
    "closing",
]
ProductClipRef = str
ProductTreatment = Literal[
    "full_focus",
    "framed_product",
    "split_explanation",
    "moving_background",
]
NarrativeSection = Literal[
    "hook",
    "problem",
    "promise",
    "product_walkthrough",
    "trust_technology",
    "editing_control",
    "finished_result",
    "closing",
]
EDITORIAL_TEMPLATE_IDS: tuple[EditorialTemplateId, ...] = (
    "hook",
    "framed_product",
    "feature_callout",
    "split_explanation",
    "proof_safety",
    "closing",
)


class EditorialTemplateCatalog(ContractModel):
    version: Literal["editorial-v1"] = "editorial-v1"
    template_ids: list[EditorialTemplateId] = Field(
        default_factory=lambda: list(EDITORIAL_TEMPLATE_IDS)
    )

    @model_validator(mode="after")
    def contains_exact_catalog(self) -> EditorialTemplateCatalog:
        if tuple(self.template_ids) != EDITORIAL_TEMPLATE_IDS:
            raise ValueError("Editorial template catalog is fixed and versioned")
        return self


class RenderSafeCrop(ContractModel):
    x: float = Field(ge=0, le=1)
    y: float = Field(ge=0, le=1)
    width: float = Field(gt=0, le=1)
    height: float = Field(gt=0, le=1)

    @model_validator(mode="after")
    def fits_frame(self) -> RenderSafeCrop:
        if self.x + self.width > 1 or self.y + self.height > 1:
            raise ValueError("Editorial crop must fit the product frame")
        return self


class EditorialSceneDraft(ContractModel):
    scene_id: NonEmptyString
    section: NarrativeSection
    template_id: EditorialTemplateId
    product_clip_ref: str = Field(pattern=r"^scene:[A-Za-z0-9._:-]+$")
    product_treatment: ProductTreatment
    start_ms: int = Field(ge=0)
    end_ms: int = Field(gt=0)
    eyebrow: str | None = Field(default=None, max_length=32)
    title: str = Field(min_length=1, max_length=80)
    body: str | None = Field(default=None, max_length=180)
    crop: RenderSafeCrop

    @model_validator(mode="after")
    def has_valid_hold(self) -> EditorialSceneDraft:
        if self.end_ms - self.start_ms < 1_000:
            raise ValueError("Editorial scenes require a one-second minimum hold")
        return self


class EditorialScene(EditorialSceneDraft):
    id: NonEmptyString


class EditorialTemplateDraft(ContractModel):
    summary: str = Field(min_length=1, max_length=240)
    scenes: list[EditorialSceneDraft] = Field(min_length=1, max_length=80)


class PresenceViolation(ContractModel):
    start_ms: int = Field(ge=0)
    end_ms: int = Field(gt=0)

    @model_validator(mode="after")
    def end_follows_start(self) -> PresenceViolation:
        if self.end_ms <= self.start_ms:
            raise ValueError("Presence violation end must follow start")
        return self


class ProductPresenceReport(ContractModel):
    duration_ms: int = Field(gt=0)
    visible_product_ms: int = Field(ge=0)
    percentage: float = Field(ge=0, le=100)
    violating_intervals: list[PresenceViolation] = Field(default_factory=list)

    @model_validator(mode="after")
    def meets_policy(self) -> ProductPresenceReport:
        expected = round(self.visible_product_ms / self.duration_ms * 100, 3)
        if abs(expected - self.percentage) > 0.01:
            raise ValueError("Product presence percentage must match visible duration")
        if self.percentage < 90:
            raise ValueError("Product footage must remain visible in at least 90% of frames")
        return self


class EditorialTemplatePlan(ContractModel):
    id: NonEmptyString
    project_id: NonEmptyString
    job_id: NonEmptyString
    run_id: NonEmptyString
    motion_plan_id: NonEmptyString
    catalog_version: Literal["editorial-v1"]
    design_version: Literal["motion-v1"]
    duration_ms: int = Field(gt=0, le=600_000)
    scene_ids: list[NonEmptyString] = Field(min_length=1, max_length=80)
    product_clip_refs: list[str] = Field(min_length=1, max_length=80)
    summary: str = Field(min_length=1, max_length=240)
    scenes: list[EditorialScene] = Field(min_length=1, max_length=80)
    created_at: datetime

    @model_validator(mode="after")
    def validates_owned_contiguous_product_scenes(self) -> EditorialTemplatePlan:
        if [scene.scene_id for scene in self.scenes] != self.scene_ids:
            raise ValueError("Editorial scenes must cover every requested scene in order")
        clips = set(self.product_clip_refs)
        if any(scene.product_clip_ref not in clips for scene in self.scenes):
            raise ValueError("Editorial scenes may use project-owned product clips only")
        expected_start = 0
        for scene in self.scenes:
            if scene.start_ms != expected_start:
                raise ValueError("Editorial scenes must be contiguous")
            expected_start = scene.end_ms
        if expected_start != self.duration_ms:
            raise ValueError("Editorial scenes must cover the complete video")
        return self

    def product_presence(self) -> ProductPresenceReport:
        visible = sum(scene.end_ms - scene.start_ms for scene in self.scenes)
        return ProductPresenceReport(
            duration_ms=self.duration_ms,
            visible_product_ms=min(visible, self.duration_ms),
            percentage=round(min(visible, self.duration_ms) / self.duration_ms * 100, 3),
            violating_intervals=[],
        )


class CompiledEditorialScene(ContractModel):
    scene_id: NonEmptyString
    template_id: EditorialTemplateId
    start_frame: int = Field(ge=0)
    end_frame: int = Field(gt=0)
    safe_margin_x: int = Field(gt=0)
    safe_margin_y: int = Field(gt=0)


class CompiledEditorialComposition(ContractModel):
    project_id: NonEmptyString
    plan_id: NonEmptyString
    width: int = Field(ge=320, le=3840)
    height: int = Field(ge=180, le=2160)
    fps: int = Field(ge=12, le=60)
    scenes: list[CompiledEditorialScene]
    product_presence: ProductPresenceReport
    deterministic_hash: str = Field(pattern=r"^[a-f0-9]{64}$")

File: packages/contracts/src/editorial.ts
import { z } from "zod";

export const editorialTemplateIdSchema = z.enum([
  "hook", "framed_product", "feature_callout", "split_explanation", "proof_safety", "closing",
]);
export const productTreatmentSchema = z.enum([
  "full_focus", "framed_product", "split_explanation", "moving_background",
]);
export const narrativeSectionSchema = z.enum([
  "hook", "problem", "promise", "product_walkthrough", "trust_technology",
  "editing_control", "finished_result", "closing",
]);
const catalog = [
  "hook", "framed_product", "feature_callout", "split_explanation", "proof_safety", "closing",
] as const;

export const editorialTemplateCatalogSchema = z.object({
  version: z.literal("editorial-v1").default("editorial-v1"),
  template_ids: z.array(editorialTemplateIdSchema).default([...catalog]),
}).strict().superRefine((value, context) => {
  if (value.template_ids.join("|") !== catalog.join("|")) {
    context.addIssue({ code: "custom", message: "Editorial template catalog is fixed" });
  }
});

export const renderSafeCropSchema = z.object({
  x: z.number().min(0).max(1), y: z.number().min(0).max(1),
  width: z.number().positive().max(1), height: z.number().positive().max(1),
}).strict().superRefine((value, context) => {
  if (value.x + value.width > 1 || value.y + value.height > 1) {
    context.addIssue({ code: "custom", message: "Editorial crop must fit the product frame" });
  }
});

const editorialSceneFields = {
  scene_id: z.string().min(1), section: narrativeSectionSchema,
  template_id: editorialTemplateIdSchema,
  product_clip_ref: z.string().regex(/^scene:[A-Za-z0-9._:-]+$/),
  product_treatment: productTreatmentSchema,
  start_ms: z.number().int().nonnegative(), end_ms: z.number().int().positive(),
  eyebrow: z.string().max(32).nullable().default(null), title: z.string().min(1).max(80),
  body: z.string().max(180).nullable().default(null), crop: renderSafeCropSchema,
};
const sceneBase = z.object(editorialSceneFields).strict();
const refineScene = (value: z.infer<typeof sceneBase>, context: z.RefinementCtx) => {
  if (value.end_ms - value.start_ms < 1_000) {
    context.addIssue({ code: "custom", message: "Editorial scenes require a one-second hold" });
  }
};
export const editorialSceneDraftSchema = sceneBase.superRefine(refineScene);
export const editorialSceneSchema = z.object({ id: z.string().min(1), ...editorialSceneFields })
  .strict().superRefine(refineScene);
export const editorialTemplateDraftSchema = z.object({
  summary: z.string().min(1).max(240), scenes: z.array(editorialSceneDraftSchema).min(1).max(80),
}).strict();

export const presenceViolationSchema = z.object({
  start_ms: z.number().int().nonnegative(), end_ms: z.number().int().positive(),
}).strict().refine((value) => value.end_ms > value.start_ms);
export const productPresenceReportSchema = z.object({
  duration_ms: z.number().int().positive(), visible_product_ms: z.number().int().nonnegative(),
  percentage: z.number().min(90).max(100),
  violating_intervals: z.array(presenceViolationSchema).default([]),
}).strict().superRefine((value, context) => {
  const expected = Math.round((value.visible_product_ms / value.duration_ms) * 100_000) / 1_000;
  if (Math.abs(expected - value.percentage) > 0.01) {
    context.addIssue({ code: "custom", message: "Product presence percentage must match duration" });
  }
});

export const editorialTemplatePlanSchema = z.object({
  id: z.string().min(1), project_id: z.string().min(1), job_id: z.string().min(1),
  run_id: z.string().min(1), motion_plan_id: z.string().min(1),
  catalog_version: z.literal("editorial-v1"), design_version: z.literal("motion-v1"),
  duration_ms: z.number().int().positive().max(600_000),
  scene_ids: z.array(z.string().min(1)).min(1).max(80),
  product_clip_refs: z.array(z.string().regex(/^scene:[A-Za-z0-9._:-]+$/)).min(1).max(80),
  summary: z.string().min(1).max(240), scenes: z.array(editorialSceneSchema).min(1).max(80),
  created_at: z.string().datetime({ offset: true }),
}).strict().superRefine((value, context) => {
  const fail = (message: string) => context.addIssue({ code: "custom", message });
  if (value.scenes.map((scene) => scene.scene_id).join("|") !== value.scene_ids.join("|")) {
    fail("Editorial scenes must cover requested scenes in order");
  }
  const clips = new Set(value.product_clip_refs);
  if (value.scenes.some((scene) => !clips.has(scene.product_clip_ref))) fail("Foreign product clip");
  let expectedStart = 0;
  for (const scene of value.scenes) {
    if (scene.start_ms !== expectedStart) fail("Editorial scenes must be contiguous");
    expectedStart = scene.end_ms;
  }
  if (expectedStart !== value.duration_ms) fail("Editorial scenes must cover the video");
});

export type EditorialTemplateId = z.infer<typeof editorialTemplateIdSchema>;
export type EditorialTemplatePlan = z.infer<typeof editorialTemplatePlanSchema>;
export type ProductPresenceReport = z.infer<typeof productPresenceReportSchema>;
